### PR TITLE
initial Aarch64 support

### PR DESCRIPTION
**Pull request by WalterBright**
Only works for a trivial function: `void foo() { }`

This is here mainly to ensure it stays in sync with the master branch as both evolve, and to not break the master.

Added undocumented `-arm` command line switch to enable it.

### CODE REVIEW

File: compiler/src/dmd/backend/x86/cgcod.d [hunk 1]
```diff
@@ -3033,19 +3038,33 @@ private extern (D)
...
-    while (i < code.length)
-    {
-        printf("%04x:", cast(int)i);
-        uint pc;
-        const sz = dmd.backend.disasm86.calccodsize(code, cast(uint)i, pc, model);
 
-        void put(char c) { printf("%c", c); }
+    void put(char c) { printf("%c", c); }
+
+    if (config.target_cpu == TARGET_Aarch64)
```
Comment (ibuclaw) [inline compiler/src/dmd/backend/x86/cgcod.d:3044]: This is the x86 backend.

File: compiler/src/dmd/backend/x86/cod3.d [hunk 2]
```diff
@@ -30,6 +30,7 @@ import dmd.backend.cc;
 import dmd.backend.cdef;
 import dmd.backend.cgcse;
 import dmd.backend.code;
+import dmd.backend.arm.cod3;
 import dmd.backend.x86.code_x86;
```
Comment (ibuclaw) [inline compiler/src/dmd/backend/x86/cod3.d:34]: I can't see you getting far mixing different architectures like this.

Comment (ibuclaw) [reply compiler/src/dmd/backend/x86/cod3.d:34]: In case this comes across as an inactionable comment, to clarify.  Changes like this make it quite clear that a lot more work needs to go into separation of IR and target modules in the back-end before the introduction of any new architecture can be done (sans this kind of spaghetti imports).

That alone would make introducing any new foreign architecture into dmd more palatable.

Comment (WalterBright) [reply compiler/src/dmd/backend/x86/cod3.d:34]: A good chunk of the code between the two will be identical. It's a balance between having them being totally separate and sharing common code. I don't know just where that balance will be, it will become apparent after I make more progress.

Comment (Herringway) [reply compiler/src/dmd/backend/x86/cod3.d:34]: Having every architecture depend on every other architecture is not sustainable. Isn't contributing a new architecture hard enough without technical debt like this?

File: compiler/src/dmd/target.d [hunk 3]
```diff
@@ -222,7 +222,11 @@ void addPredefinedGlobalIdentifiers(const ref Target tgt)
     addCRuntimePredefinedGlobalIdent(tgt.c);
     addCppRuntimePredefinedGlobalIdent(tgt.cpp);
 
-    if (tgt.isX86_64)
+    if (tgt.isAarch64)
+    {
+        VersionCondition.addPredefinedGlobalIdent("Aarch64");
```
Comment (dd86k) [inline compiler/src/dmd/target.d:227]: Documentation mentions `AArch64`, typo?

Comment (WalterBright) [reply compiler/src/dmd/target.d:227]: which documentation?

Comment (dd86k) [reply compiler/src/dmd/target.d:227]: Sorry, predefined versions, as Iain says.

Comment (WalterBright) [reply compiler/src/dmd/target.d:227]: Ok. I'll fix it.

Comment (WalterBright) [reply compiler/src/dmd/target.d]: done

File: compiler/src/dmd/backend/x86/cod3.d (hunk 2 above)
Comment (WalterBright) [reply compiler/src/dmd/backend/x86/cod3.d:34]: It won't be clear where the separation should be until I make much more progress.

File: compiler/src/dmd/backend/debugprint.d [hunk 4]
```diff
@@ -331,7 +331,7 @@ void WRdefnod()
 @trusted
 const(char)* fl_str(FL fl)
 {
-    __gshared const char*[FLMAX] fls =
```
Comment (thewilsonator) [inline compiler/src/dmd/backend/debugprint.d:334]: please split this trivial change into a separate PR

Comment (thewilsonator) [reply compiler/src/dmd/backend/debugprint.d:334]: https://github.com/dlang/dmd/pull/16665

Comment (WalterBright) [reply compiler/src/dmd/backend/debugprint.d:334]: Thanks, Nick!

File: compiler/src/dmd/target.d [hunk 5]
```diff
@@ -360,6 +365,7 @@ extern (C++) struct Target
     /// Architecture name
     const(char)[] architectureName;
     CPU cpu;                // CPU instruction set to target
+    bool isAArch64;         // generate 64 bit Arm code
```
Comment (thewilsonator) [inline compiler/src/dmd/target.d:368]: https://github.com/dlang/dmd/pull/16669

**Comment by WalterBright** [comment]
Added changelog entry

**Comment by WalterBright** [comment]
@dkorpel ,
```
int main() { return 0; }
```
now compiles, links, and runs on a Raspberry Pi! Time to merge!

File: changelog/dmd.AArch64.dd [hunk 6]
```diff
@@ -0,0 +1,28 @@
+# Add code generation for AArch64 CPU
+
+Adds the switch `-arm` which causes dmd to emit AArch64 code.
```
Comment (dkorpel) [inline changelog/dmd.AArch64.dd:3]: As we agreed on in the meeting Friday, the switch should not be exposed to users yet.

**Review by dkorpel — CHANGES_REQUESTED** [review]

File: changelog/dmd.AArch64.dd (hunk 6 above)
Comment (WalterBright) [reply changelog/dmd.AArch64.dd:3]: It remains undocumented as far as users are concerned.

Nobody can help with the development of it if they cannot turn it on.

Besides, it's a nuisance and impractical to keep altering the source code every time a push is done. Blech.

Comment (dkorpel) [reply changelog/dmd.AArch64.dd:3]: The backend code can still get merged, but you agreed to Rikki's comment that the switch shouldn't be public remember? Advertising the `-arm` switch in the changelog is the opposite of that.

Comment (WalterBright) [reply changelog/dmd.AArch64.dd:3]: ping @dkorpel

Comment (dkorpel) [reply changelog/dmd.AArch64.dd:3]: Still unaddressed

Comment (WalterBright) [reply changelog/dmd.AArch64.dd:3]: I'm going to leave the -arm switch in, and leave it undocumented, for reasons explained above.

Comment (WalterBright) [reply changelog/dmd.AArch64.dd:3]: @thewilsonator asked for a changelog.

Comment (WalterBright) [reply changelog/dmd.AArch64.dd:3]: documenting it would be putting it on the command line documentation, both what the compiler emits and the web page for it.

Comment (Herringway) [reply changelog/dmd.AArch64.dd:3]: We users read changelogs. It's important to keep up with what's changing in the tools you use. DMD's changelogs would be a lot more useful to us if they weren't treated like to-do lists (D has an issue tracker for that kind of thing, doesn't it?) littered with unfinished language features and compiler switches.

Comment (dkorpel) [reply changelog/dmd.AArch64.dd:3]: Do you understand it's frustrating when in a DLF meeting we conclude "we're going to do X", then the next day you're doing Y instead, and refuse to address or even acknowledge this disrepancy?

Comment (WalterBright) [reply changelog/dmd.AArch64.dd:3]: It remains undocumented in the compiler's list of switches and the documentation for how to use the compiler. It's entirely proper to have a changelog entry for a 6000 line addition.

**Comment by WalterBright** [comment]
Aaaannnndddd - **hello world!** compiles and runs now!

**Comment by WalterBright** [comment]
This commit changed nothing in the x86 code generation, so why is it failing?

**Comment by WalterBright** [comment]
Pushing it again seems to have "fixed" it.

**Comment by WalterBright** [comment]
How does disasm86.d, with its last commit 2 weeks ago, suddenly start failing the LDC Ubuntu run of its unittests?

https://github.com/dlang/dmd/commit/a74614204e7e0fca0df7c61cc020443259dd3b14

```
core.exception.AssertError@src/dmd/backend/x86/disasm86.d(3760): Assertion failure
```

**Comment by WalterBright** [comment]
The LDC Ubuntu test is now working again, though I still have made no changes in that direction.

**Comment by WalterBright** [comment]
The LDC Ubuntu is back to failing again.

**Review by dkorpel — APPROVED** [review]

**Comment by tgehr** [comment]
In target.d:

> 426 | isX86 = !isX86_64;

Looks wrong.

**Comment by WalterBright** [comment]
@tgehr the process of setting those variables is an ugly tangle that I had only partially straightened out. I added an assert that it was set up correctly.

**Comment by WalterBright** [comment]
P.S. I fixed the heisenbug in the disasm86 code.

**Comment by tgehr** [comment]
> @tgehr the process of setting those variables is an ugly tangle that I had only partially straightened out. I added an assert that it was set up correctly.

```d
isX86 = !isX86_64;
assert(isX86 + isX86_64 + isAArch64 == 1); // there can be only one
```

This assertion is equivalent to `assert(!isAArch64)`.
